### PR TITLE
[TTAHUB-1317] Only find matching objectives based on trim

### PR DESCRIPTION
**Pull request by AdamAdHocTeam**
## Description of change

This solves a bug that I introduced as part of the fix for [TTAHUB-1240](https://github.com/HHS/Head-Start-TTADP/pull/1204/files).

In attempt to prevent showing duplicate objectives with leading or trailing spaces we perform a trim on the returned objective title. We cannot do this as it flags the title as changed in the Sequelize instance. This causes a validation error to be thrown if the goal is on an approved report (user case).

For now we will see if a non-trailing version of the objective title exists. Else we will return the title as is with spaces.

## How to test

You can edit recipient 'Blueprints' for G-20778, G-12658 and attempt to change all statuses to complete and save.

OR

Create a goal with a few objectives on an approved report. Make sure one of the Objective titles has trailing spaces. You should still be able to edit the goal and make certain changes to the Objective and save without error.

## Issue(s)

* https://ocio-jira.acf.hhs.gov/browse/TTAHUB-1317

### Every PR

- [x] Meets issue criteria
- [x] JIRA ticket status updated
- [x] Code is meaningfully tested
- [x] Meets accessibility standards (WCAG 2.1 Levels A, AA)
- [ ] API Documentation updated
- [ ] Boundary diagram updated
- [ ] Logical Data Model updated
- [ ] [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions
- [ ] UI review complete

### Production Deploy

- [ ] Staging smoke test completed

### After merge/deploy

- [x] Update JIRA ticket status

### CODE REVIEW

**Comment by thewatermethod** [comment]
With this change, objectives are not de-duplicated on the "edit goals" page (when you click "edit" from the RTR) or in the objective select dropdown in the Activity Report. I confirmed that they were correctly de-duplicated on main

Example: /recipient-tta-records/706/region/1/goals?id[]=27525,27528

**Comment by AdamAdHocTeam** [comment]
> 27528

The issue here is that both objectives titles have trailing spaces. So when we attempt to find the existing objective, we need to trim both titles.

This should be ready for another look.

**Comment by thewatermethod** [comment]
Found a couple of bugs. I know your code isn't responsible for introducing them but I wonder if they should be fixed before we merge this fix in since it will open us up to putting data in a weird state.

Editing the same objectives I linked to before
- updating the status from the RTR only updates _one_ of the objective statuses
- updating the status from the AR does not seem to update either objective at all, only the ActivityReportObjective

**Comment by thewatermethod** [comment]
Regardless, up to you - if you want me to approve this to unblock a user, maybe you could write up some jira tickets to document these bugs.

**Comment by AdamAdHocTeam** [comment]
> Found a couple of bugs. I know your code isn't responsible for introducing them but I wonder if they should be fixed before we merge this fix in since it will open us up to putting data in a weird state.
> 
> Editing the same objectives I linked to before
> 
> * updating the status from the RTR only updates _one_ of the objective statuses
> * updating the status from the AR does not seem to update either objective at all, only the ActivityReportObjective (I added the objective to an AR, changed the status, and approved the report)

Lets go over these once your back. While testing I'm getting some mixed results (maybe I'm using the wrong data).

File: src/models/hooks/activityReport.js
Comment (GarrettEHill) [inline src/models/hooks/activityReport.js:233]: is this happening anywhere else?

Comment (GarrettEHill) [reply src/models/hooks/activityReport.js:233]: Looks like maybe a few:
./src/services/recipient.js:204:        endDate: report.endDate > a.endDate ? report.endDate : a.endDate,
./src/services/recipient.js:253:    a.endDate === b.endDate ? a.id < b.id
./src/services/recipient.js:254:      : a.endDate < b.endDate) ? 1 : -1));
./src/widgets/totalHrsAndRecipientGraph.js:129:      if (!arDates.find((cache) => cache.id === r.id && cache.date === r.startDate)) {

Comment (AdamAdHocTeam) [reply src/models/hooks/activityReport.js:233]: This was the test I ran.

// Check === with Dates
const dateTestE = '01/05/2023';
const dateELess = '01/05/2023';
const checkE = new Date(dateTestE) === new Date(dateELess);
console.log('ResultE: ', checkE);

// Check === with Strings
const dateTestF = '01/05/2023';
const dateFLess = '01/05/2023';
const checkF = dateTestF === dateFLess;
console.log('ResultF: ', checkF);

**Comment by AdamAdHocTeam** [comment]
> > Found a couple of bugs. I know your code isn't responsible for introducing them but I wonder if they should be fixed before we merge this fix in since it will open us up to putting data in a weird state.
> > Editing the same objectives I linked to before
> > 
> > * updating the status from the RTR only updates _one_ of the objective statuses
> > * updating the status from the AR does not seem to update either objective at all, only the ActivityReportObjective (I added the objective to an AR, changed the status, and approved the report)
> 
> Lets go over these once your back. While testing I'm getting some mixed results (maybe I'm using the wrong data).

@thewatermethod I have fixed the two additional issues you mentioned in your comments.
1. The first issue will be fixed in ttahub-1320-update-all-grants.
2. The second issue is included in this PR (approved reports).

**Comment by thewatermethod** [comment]
@AdamAdHocTeam Is there any way we could change our tests to catch some of the cases we've turned up during the review? Specifically, can we assert via a test somewhere:

- that it's possible to change the status of a "rolled-up" objective (the original bug) 
- that only one objective will be returned by the backend even if one has a trailing space and one doesn't
- that approving an AR will change an objective's status, even if that objective is on another AR

**Comment by AdamAdHocTeam** [comment]
> * that it's possible to change the status of a "rolled-up" objective (the original bug)

"that it's possible to change the status of a "rolled-up" objective (the original bug)"

This bug is fixed by another PR.

**Comment by AdamAdHocTeam** [comment]
> @AdamAdHocTeam Is there any way we could change our tests to catch some of the cases we've turned up during the review? Specifically, can we assert via a test somewhere:
> 
> * that it's possible to change the status of a "rolled-up" objective (the original bug)
> * that only one objective will be returned by the backend even if one has a trailing space and one doesn't
> * that approving an AR will change an objective's status, even if that objective is on another AR

@thewatermethod add requested test changes.

**Review by thewatermethod — APPROVED** [review]
Nice work, thanks for adding those tests

**Pull request merged**
